src/test/core/JobCounter_test.cpp

```diff
@@ -0,0 +1,124 @@
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/core/JobCounter.h>
+#include <ripple/beast/unit_test.h>
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+namespace ripple {
+
+//------------------------------------------------------------------------------
+
+class JobCounter_test : public beast::unit_test::suite
+{
+    void testWrap()
+    {
+        // Verify reference counting.
+        JobCounter jobCounter;
+        BEAST_EXPECT (jobCounter.count() == 0);
+        {
+            auto wrapped1 = jobCounter.wrap ([] (Job&) {});
+            BEAST_EXPECT (jobCounter.count() == 1);
+
+            // wrapped1 should be callable with a Job.
+            {
+                Job j;
+                (*wrapped1)(j);
+            }
+            {
+                // Copy should increase reference count.
+                auto wrapped2 (wrapped1);
+                BEAST_EXPECT (jobCounter.count() == 2);
+                {
+                    // Moving from wrapped2 doesn't decrease count.
+                    auto wrapped3 (std::move(wrapped2));
+                    BEAST_EXPECT (jobCounter.count() == 3);
+                    {
+                        // An additional Job also increases count.
+                        auto wrapped4 = jobCounter.wrap ([] (Job&) {});
+                        BEAST_EXPECT (jobCounter.count() == 4);
+                    }
+                    BEAST_EXPECT (jobCounter.count() == 3);
+                }
+                BEAST_EXPECT (jobCounter.count() == 2);
+            }
+            BEAST_EXPECT (jobCounter.count() == 1);
+        }
+        BEAST_EXPECT (jobCounter.count() == 0);
+
+        // Join with 0 count should not stall.
+        jobCounter.join();
+
+        // Wrapping a Job after join() should return boost::none.
+        BEAST_EXPECT (jobCounter.wrap ([] (Job&) {}) == boost::none);
+    }
+
+    void testWaitOnJoin()
+    {
+        // Verify reference counting.
+        JobCounter jobCounter;
+        BEAST_EXPECT (jobCounter.count() == 0);
+
+        auto job = (jobCounter.wrap ([] (Job&) {}));
+        BEAST_EXPECT (jobCounter.count() == 1);
+
+        // Calling join() now should stall, so do it on a different thread.
+        std::atomic<bool> threadActive {false};
+        std::thread localThread ([&jobCounter, &threadActive] ()
+        {
+            threadActive.store (true);
+
+            // Should stall after calling join.
+            jobCounter.join();
+
+            threadActive.store (false);
+        });
+
+        // Wait for the thread to start.
+        while (threadActive.load() == false);
+        BEAST_EXPECT (jobCounter.count() == 1);
+
+        // Wait a millisecond.  The thread should still be active.
+        using namespace std::chrono_literals;
+        std::this_thread::sleep_for (1ms);
```

Comment:
I actually prefer the sleep here.  What I'm doing is developing some comfort that `localThread` really is stalled inside the `join()` by waiting one millisecond.  So the wait is deterministic.  I can't think of a way to develop a similar confidence using a `condition_variable`.